File: Object-C/BlackFireIM/Example/Lib/MSIMSDK-UIKit/VoiceKit/MSVideoCallView.m
//
//  MSVideoCallView.m
//  BlackFireIM
//
//  Created by benny wang on 2021/7/23.
//

#import "MSVideoCallView.h"
#import "MSIMSDK-UIKit.h"
#import <MSIMSDK/MSIMSDK.h>
#import <SDWebImage.h>


@implementation MSVideoCallView

- (instancetype)initWithFrame:(CGRect)frame
{
    self = [super initWithFrame:frame];
    if (self) {
        self.localView = [[UIView alloc]initWithFrame:UIScreen.mainScreen.bounds];
        [self addSubview:self.localView];
        
        self.remoteView = [[UIView alloc]initWithFrame:CGRectMake(Screen_Width-20-90, StatusBar_Height + NavBar_Height, 90, 160)];
        self.remoteView.layer.cornerRadius = 6;
        self.remoteView.clipsToBounds = YES;
        [self addSubview:self.remoteView];
        UITapGestureRecognizer *remoteTap = [[UITapGestureRecognizer alloc]initWithTarget:self action:@selector(remoteViewTap)];
        [self.remoteView addGestureRecognizer:remoteTap];
        
        self.avatarIcon = [[UIImageView alloc]initWithFrame:CGRectMake(20, StatusBar_Height + NavBar_Height, 70, 70)];
        self.avatarIcon.layer.cornerRadius = 4;
        self.avatarIcon.layer.masksToBounds = YES;
        self.avatarIcon.contentMode = UIViewContentModeScaleAspectFill;
        [self addSubview:self.avatarIcon];
        
        self.durationL = [[UILabel alloc]initWithFrame:CGRectMake(Screen_Width*0.5-50, StatusBar_Height + NavBar_Height, 100, 20)];
        self.durationL.textColor = [UIColor whiteColor];
        self.durationL.font = [UIFont systemFontOfSize:16];
        self.durationL.textAlignment = NSTextAlignmentCenter;
        [self addSubview:self.durationL];
        
        self.nickNamekL = [[UILabel alloc]initWithFrame:CGRectMake(self.avatarIcon.maxX+15, self.avatarIcon.y, 200, 35)];
        self.nickNamekL.font = [UIFont boldSystemFontOfSize:20];
        self.nickNamekL.textColor = [UIColor whiteColor];
        [self addSubview:self.nickNamekL];
        
        self.noticeL = [[UILabel alloc]initWithFrame:CGRectMake(self.nickNamekL.x, self.nickNamekL.maxY+12, 200, 20)];
        self.noticeL.font = [UIFont systemFontOfSize:14];
        self.noticeL.textColor = [UIColor whiteColor];
        [self addSubview:self.noticeL];
        
        self.cancelBtn = [UIButton buttonWithType:UIButtonTypeCustom];
        [self.cancelBtn setTitle:TUILocalizableString(Cancel) forState:UIControlStateNormal];
        [self.cancelBtn setTitleColor:[UIColor whiteColor] forState:UIControlStateNormal];
        self.cancelBtn.titleLabel.font = [UIFont systemFontOfSize:12];
        [self.cancelBtn setImage:[UIImage imageNamed:TUIKitResource(@"ic_hangup")] forState:UIControlStateNormal];
        [self.cancelBtn addTarget:self action:@selector(cancelBtnDidClick:) forControlEvents:UIControlEventTouchUpInside];
        self.cancelBtn.frame = CGRectMake(Screen_Width*0.5-42, Screen_Height-Bottom_SafeHeight-30-114, 84, 114);
        [self addSubview:self.cancelBtn];
        [self.cancelBtn verticalImageAndTitle:15];
        
        self.hangupBtn = [UIButton buttonWithType:UIButtonTypeCustom];
        [self.hangupBtn setTitle:TUILocalizableString(TUIKitCallCancel) forState:UIControlStateNormal];
        [self.hangupBtn setTitleColor:[UIColor whiteColor] forState:UIControlStateNormal];
        self.hangupBtn.titleLabel.font = [UIFont systemFontOfSize:12];
        [self.hangupBtn setImage:[UIImage imageNamed:TUIKitResource(@"ic_hangup")] forState:UIControlStateNormal];
        [self.hangupBtn addTarget:self action:@selector(hangupBtnDidClick:) forControlEvents:UIControlEventTouchUpInside];
        self.hangupBtn.frame = CGRectMake(Screen_Width*0.5-self.cancelBtn.width-30, self.cancelBtn.y, self.cancelBtn.width, self.cancelBtn.height);
        [self addSubview:self.hangupBtn];
        [self.hangupBtn verticalImageAndTitle:15];
        
        self.cameraBtn = [UIButton buttonWithType:UIButtonTypeCustom];
        [self.cameraBtn setTitle:TUILocalizableString(TUIKitCallCameraSwitch) forState:UIControlStateNormal];
        [self.cameraBtn setTitleColor:[UIColor whiteColor] forState:UIControlStateNormal];
        self.cameraBtn.titleLabel.font = [UIFont systemFontOfSize:12];
        [self.cameraBtn setImage:[UIImage imageNamed:TUIKitResource(@"ic_camera")] forState:UIControlStateNormal];
        [self.cameraBtn addTarget:self action:@selector(cameraBtnDidClick:) forControlEvents:UIControlEventTouchUpInside];
        self.cameraBtn.frame = CGRectMake(Screen_Width*0.5+30, self.hangupBtn.y, self.cancelBtn.width, self.cancelBtn.height);
        [self addSubview:self.cameraBtn];
        [self.cameraBtn verticalImageAndTitle:15];
        
        self.rejectBtn = [UIButton buttonWithType:UIButtonTypeCustom];
        [self.rejectBtn setTitle:TUILocalizableString(TUIKitCallReject) forState:UIControlStateNormal];
        [self.rejectBtn setTitleColor:[UIColor whiteColor] forState:UIControlStateNormal];
        self.rejectBtn.titleLabel.font = [UIFont systemFontOfSize:12];
        [self.rejectBtn setImage:[UIImage imageNamed:TUIKitResource(@"ic_hangup")] forState:UIControlStateNormal];
        [self.rejectBtn addTarget:self action:@selector(rejectBtnDidClick:) forControlEvents:UIControlEventTouchUpInside];
        self.rejectBtn.frame = CGRectMake(Screen_Width*0.5-self.cancelBtn.width-30, self.cancelBtn.y, self.cancelBtn.width, self.cancelBtn.height);
        [self addSubview:self.rejectBtn];
        [self.rejectBtn verticalImageAndTitle:15];

        self.acceptBtn = [UIButton buttonWithType:UIButtonTypeCustom];
        [self.acceptBtn setTitle:TUILocalizableString(TUIKitCallAccept) forState:UIControlStateNormal];
        [self.acceptBtn setTitleColor:[UIColor whiteColor] forState:UIControlStateNormal];
        self.acceptBtn.titleLabel.font = [UIFont systemFontOfSize:12];
        [self.acceptBtn setImage:[UIImage imageNamed:TUIKitResource(@"ic_dialing")] forState:UIControlStateNormal];
        [self.acceptBtn addTarget:self action:@selector(acceptBtnDidClick:) forControlEvents:UIControlEventTouchUpInside];
        self.acceptBtn.frame = CGRectMake(Screen_Width*0.5+30, self.rejectBtn.y, self.cancelBtn.width, self.cancelBtn.height);
        [self addSubview:self.acceptBtn];
        [self.acceptBtn verticalImageAndTitle:15];
    }
    return self;
}

- (void)initDataWithSponsor:(BOOL)isMe partner_id:(NSString *)partner_id
{
    [[MSProfileProvider provider] providerProfile:partner_id complete:^(MSProfileInfo * _Nullable profile) {
        if (profile) {
            [self.avatarIcon sd_setImageWithURL:[NSURL URLWithString:profile.avatar]];
            self.nickNamekL.text = profile.nick_name;
        }
    }];
    self.noticeL.text = isMe ? TUILocalizableString(TUIKitCallWaitingForAccept) : TUILocalizableString(TUIKitCallInviteYouVideoCall);
    self.cancelBtn.hidden = !isMe;
    self.cameraBtn.hidden = YES;
    self.rejectBtn.hidden = isMe;
    self.acceptBtn.hidden = isMe;
    self.durationL.hidden = YES;
    self.hangupBtn.hidden = YES;
    self.remoteView.hidden = YES;
}

#pragma mark - btn event

- (void)remoteViewTap
{
    if ([self.delegate respondsToSelector:@selector(video_remoteViewDidTap)]) {
        [self.delegate video_remoteViewDidTap];
    }
}

- (void)cancelBtnDidClick:(UIButton *)sender
{
    if ([self.delegate respondsToSelector:@selector(video_cancelBtnDidClick)]) {
        [self.delegate video_cancelBtnDidClick];
    }
}

- (void)cameraBtnDidClick:(UIButton *)sender
{
    if ([self.delegate respondsToSelector:@selector(video_cameraBtnDidClick)]) {
        [self.delegate video_cameraBtnDidClick];
    }
}

- (void)rejectBtnDidClick:(UIButton *)sender
{
    if ([self.delegate respondsToSelector:@selector(video_rejectBtnDidClick)]) {
        [self.delegate video_rejectBtnDidClick];
    }
}

- (void)acceptBtnDidClick:(UIButton *)sender
{
    if ([self.delegate respondsToSelector:@selector(video_acceptBtnDidClick)]) {
        [self.delegate video_acceptBtnDidClick];
    }
}

- (void)hangupBtnDidClick:(UIButton *)sender
{
    if ([self.delegate respondsToSelector:@selector(video_hangupBtnDidClick)]) {
        [self.delegate video_hangupBtnDidClick];
    }
}

@end

File: Swift/MSIM-Swift/Lib/MSIMSDK-UIKit/CallKit/MSVideoCallView.swift
//
//  MSVideoCallView.swift
//  MSIM-Swift
//
//  Created by benny wang on 2021/9/28.
//

import UIKit
import MSIMSDK
import Kingfisher

protocol MSVideoCallViewDelegate: NSObjectProtocol {
    
    func video_remoteViewDidTap()
    func video_cancelBtnDidClick()
    func video_cameraBtnDidClick()
    func video_rejectBtnDidClick()
    func video_acceptBtnDidClick()
    func video_hangupBtnDidClick()
}
class MSVideoCallView: UIView {

    weak var delegate: MSVideoCallViewDelegate?
    
    lazy var localView: UIView = {
        let localView = UIView(frame: UIScreen.main.bounds)
        return localView
    }()

    lazy var remoteView: UIView = {
        let remoteView = UIView(frame: CGRect(x: UIScreen.width - 20 - 90, y: UIScreen.status_navi_height, width: 90, height: 160))
        remoteView.layer.cornerRadius = 6
        remoteView.clipsToBounds = true
        return remoteView
    }()
    
    lazy var avatarIcon: UIImageView = {
        let avatarIcon = UIImageView(frame: CGRect(x: 20, y: UIScreen.status_navi_height, width: 70, height: 70))
        avatarIcon.layer.cornerRadius = 4
        avatarIcon.layer.masksToBounds = true
        avatarIcon.contentMode = .scaleAspectFill
        return avatarIcon
    }()
    
    lazy var durationL: UILabel = {
        let durationL = UILabel(frame: CGRect(x: UIScreen.width * 0.5 - 50, y: UIScreen.status_navi_height, width: 100, height: 20))
        durationL.textColor = .white
        durationL.font = .systemFont(ofSize: 16)
        durationL.textAlignment = .center
        return durationL
    }()
    
    lazy var nickNameL: UILabel = {
        let nickNameL = UILabel(frame: CGRect(x: avatarIcon.right + 15, y: avatarIcon.top, width: 200, height: 35))
        nickNameL.font = .systemFont(ofSize: 20)
        nickNameL.textColor = .white
        nickNameL.textAlignment = .center
        return nickNameL
    }()
    
    lazy var noticeL: UILabel = {
        let noticeL = UILabel(frame: CGRect(x: nickNameL.left, y: nickNameL.bottom + 12, width: 200, height: 20))
        noticeL.font = .systemFont(ofSize: 14)
        noticeL.textColor = .white
        noticeL.textAlignment = .center
        return noticeL
    }()
    
    lazy var cancelBtn: UIButton = {
        let cancelBtn = UIButton(type: .custom)
        cancelBtn.setTitle(Bundle.bf_localizedString(key: "Cancel"), for: .normal)
        cancelBtn.setTitleColor(.white, for: .normal)
        cancelBtn.titleLabel?.font = .systemFont(ofSize: 12)
        cancelBtn.setImage(UIImage.bf_imageNamed(name: "ic_hangup"), for: .normal)
        cancelBtn.addTarget(self, action: #selector(cancelBtnDidClick), for: .touchUpInside)
        cancelBtn.frame = CGRect(x: UIScreen.width * 0.5 - 42, y: UIScreen.height - UIScreen.safeAreaBottomHeight - 30 - 114, width: 84, height: 114)
        return cancelBtn
    }()
    
    lazy var hangupBtn: UIButton = {
        let hangupBtn = UIButton(type: .custom)
        hangupBtn.setTitle(Bundle.bf_localizedString(key: "TUIKitCallCancel"), for: .normal)
        hangupBtn.setTitleColor(.white, for: .normal)
        hangupBtn.titleLabel?.font = .systemFont(ofSize: 12)
        hangupBtn.setImage(UIImage.bf_imageNamed(name: "ic_hangup"), for: .normal)
        hangupBtn.addTarget(self, action: #selector(hangupBtnDidClick), for: .touchUpInside)
        hangupBtn.frame = CGRect(x: UIScreen.width * 0.5 - cancelBtn.width - 30, y: cancelBtn.top, width: cancelBtn.width, height: cancelBtn.height)
        return hangupBtn
    }()
    
    lazy var cameraBtn: UIButton = {
        let cameraBtn = UIButton(type: .custom)
        cameraBtn.setTitle(Bundle.bf_localizedString(key: "TUIKitCallCameraSwitch"), for: .normal)
        cameraBtn.setTitleColor(.white, for: .normal)
        cameraBtn.titleLabel?.font = .systemFont(ofSize: 12)
        cameraBtn.setImage(UIImage.bf_imageNamed(name: "ic_camera"), for: .normal)
        cameraBtn.addTarget(self, action: #selector(cameraBtnDidClick), for: .touchUpInside)
        cameraBtn.frame = CGRect(x: UIScreen.width * 0.5 + 30, y: hangupBtn.top, width: cancelBtn.width, height: cancelBtn.height)
        return cameraBtn
    }()
    
    lazy var rejectBtn: UIButton = {
        let rejectBtn = UIButton(type: .custom)
        rejectBtn.setTitle(Bundle.bf_localizedString(key: "TUIKitCallReject"), for: .normal)
        rejectBtn.setTitleColor(.white, for: .normal)
        rejectBtn.titleLabel?.font = .systemFont(ofSize: 12)
        rejectBtn.setImage(UIImage.bf_imageNamed(name: "ic_hangup"), for: .normal)
        rejectBtn.addTarget(self, action: #selector(rejectBtnDidClick), for: .touchUpInside)
        rejectBtn.frame = CGRect(x: UIScreen.width * 0.5 - 30 - cancelBtn.width, y: cancelBtn.top, width: cancelBtn.width, height: cancelBtn.height)
        return rejectBtn
    }()
    
    lazy var acceptBtn: UIButton = {
        let acceptBtn = UIButton(type: .custom)
        acceptBtn.setTitle(Bundle.bf_localizedString(key: "TUIKitCallAccept"), for: .normal)
        acceptBtn.setTitleColor(.white, for: .normal)
        acceptBtn.titleLabel?.font = .systemFont(ofSize: 12)
        acceptBtn.setImage(UIImage.bf_imageNamed(name: "ic_dialing"), for: .normal)
        acceptBtn.addTarget(self, action: #selector(acceptBtnDidClick), for: .touchUpInside)
        acceptBtn.frame = CGRect(x: UIScreen.width * 0.5 + 30, y: rejectBtn.top, width: cancelBtn.width, height: cancelBtn.height)
        return acceptBtn
    }()
    
    override init(frame: CGRect) {
        super.init(frame: frame)
        addSubview(localView)
        addSubview(remoteView)
        
        let remoteTap = UITapGestureRecognizer(target: self, action: #selector(remoteViewTap))
        remoteView.addGestureRecognizer(remoteTap)
        
        addSubview(avatarIcon)
        addSubview(durationL)
        addSubview(nickNameL)
        addSubview(noticeL)
        addSubview(cancelBtn)
        cancelBtn.verticalImageAndTitle(spacing: 15)
        addSubview(hangupBtn)
        hangupBtn.verticalImageAndTitle(spacing: 15)
        addSubview(cameraBtn)
        cameraBtn.verticalImageAndTitle(spacing: 15)
        addSubview(rejectBtn)
        rejectBtn.verticalImageAndTitle(spacing: 15)
        addSubview(acceptBtn)
        acceptBtn.verticalImageAndTitle(spacing: 15)
    }
    
    func initDataWithSponsor(isMe: Bool,partner_id: String) {
        
        MSProfileProvider.shared().providerProfile(partner_id) { profile in
            if profile != nil {
                self.avatarIcon.kf.setImage(with: URL(string: profile!.avatar))
                self.nickNameL.text = profile?.nick_name
            }
        }
        noticeL.text = isMe ? Bundle.bf_localizedString(key: "TUIKitCallWaitingForAccept") : Bundle.bf_localizedString(key: "TUIKitCallInviteYouVideoCall")
        cancelBtn.isHidden = !isMe
        cameraBtn.isHidden = true
        rejectBtn.isHidden = isMe
        acceptBtn.isHidden = isMe
        durationL.isHidden = true
        hangupBtn.isHidden = true
        remoteView.isHidden = true
    }
    
    @objc func remoteViewTap() {
        self.delegate?.video_remoteViewDidTap()
    }
    
    @objc func cancelBtnDidClick() {
        self.delegate?.video_cancelBtnDidClick()
    }
    
    @objc func cameraBtnDidClick() {
        self.delegate?.video_cameraBtnDidClick()
    }
    
    @objc func rejectBtnDidClick() {
        self.delegate?.video_rejectBtnDidClick()
    }
    
    @objc func acceptBtnDidClick() {
        self.delegate?.video_acceptBtnDidClick()
    }
    
    @objc func hangupBtnDidClick() {
        self.delegate?.video_hangupBtnDidClick()
    }
    
    required init?(coder: NSCoder) {
        fatalError("init(coder:) has not been implemented")
    }

}
